File: core/src/main/java/com/emarsys/core/database/repository/AbstractSqliteRepository.java
package com.emarsys.core.database.repository;

import android.content.ContentValues;
import android.database.Cursor;

import com.emarsys.core.database.CoreSQLiteDatabase;
import com.emarsys.core.database.helper.DbHelper;
import com.emarsys.core.util.Assert;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractSqliteRepository<T> implements Repository<T, SqlSpecification> {

    String tableName;
    DbHelper dbHelper;

    public AbstractSqliteRepository(String tableName, DbHelper dbHelper) {
        Assert.notNull(dbHelper, "DbHelper must not be null!");
        Assert.notNull(tableName, "TableName must not be null!");
        this.tableName = tableName;
        this.dbHelper = dbHelper;
    }

    protected abstract ContentValues contentValuesFromItem(T item);

    protected abstract T itemFromCursor(Cursor cursor);

    @Override
    public void add(T item) {
        Assert.notNull(item, "Item must not be null!");

        ContentValues contentValues = contentValuesFromItem(item);

        CoreSQLiteDatabase database = dbHelper.getWritableCoreDatabase();
        database.beginTransaction();
        database.insert(tableName, null, contentValues);
        database.setTransactionSuccessful();
        database.endTransaction();
    }

    @Override
    public List<T> query(SqlSpecification specification) {
        Assert.notNull(specification, "Specification must not be null!");

        CoreSQLiteDatabase database = dbHelper.getReadableCoreDatabase();
        Cursor cursor = database.query(
                specification.isDistinct(),
                tableName,
                specification.getColumns(),
                specification.getSelection(),
                specification.getSelectionArgs(),
                specification.getGroupBy(),
                specification.getHaving(),
                specification.getOrderBy(),
                specification.getLimit()
        );

        return mapCursorToResultList(cursor);
    }

    @Override
    public void remove(SqlSpecification specification) {
        Assert.notNull(specification, "Specification must not be null!");

        CoreSQLiteDatabase database = dbHelper.getWritableCoreDatabase();
        database.beginTransaction();
        database.delete(
                tableName,
                specification.getSelection(),
                specification.getSelectionArgs());
        database.setTransactionSuccessful();
        database.endTransaction();
    }

    @Override
    public boolean isEmpty() {
        CoreSQLiteDatabase database = dbHelper.getReadableCoreDatabase();
        Cursor cursor = database.rawQuery(
                String.format("SELECT COUNT(*) FROM %s;", tableName),
                null);
        cursor.moveToFirst();
        int count = cursor.getInt(cursor.getColumnIndex("COUNT(*)"));
        cursor.close();
        return count == 0;
    }

    private List<T> mapCursorToResultList(Cursor cursor) {
        List<T> result = new ArrayList<>();

        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast()) {
                T item = itemFromCursor(cursor);
                result.add(item);
                cursor.moveToNext();
            }
        }
        cursor.close();

        return result;
    }
}

File: core/src/main/java/com/emarsys/core/database/repository/AbstractSqliteRepository.kt
package com.emarsys.core.database.repository

import android.content.ContentValues
import android.database.Cursor
import com.emarsys.core.database.CoreSQLiteDatabase
import com.emarsys.core.database.helper.DbHelper
import com.emarsys.core.handler.ConcurrentHandlerHolder


private inline fun <T> CoreSQLiteDatabase.inTransaction(statement: CoreSQLiteDatabase.() -> T): T {
    this.beginTransaction()
    val result: T
    try {
        result = statement(this)
        this.setTransactionSuccessful()
    } finally {
        this.endTransaction()
    }
    return result
}

abstract class AbstractSqliteRepository<T>(
    var tableName: String,
    var dbHelper: DbHelper,
    var concurrentHandlerHolder: ConcurrentHandlerHolder
) : Repository<T, SqlSpecification> {

    abstract fun contentValuesFromItem(item: T): ContentValues
    abstract fun itemFromCursor(cursor: Cursor): T

    override fun add(item: T) {
        val contentValues = contentValuesFromItem(item)
        val database = dbHelper.writableCoreDatabase
        database.inTransaction {
            insert(tableName, null, contentValues)
        }
    }

    override fun update(item: T, specification: SqlSpecification): Int {
        val values = contentValuesFromItem(item)
        val database = dbHelper.writableCoreDatabase
        var result = 0

        database.inTransaction {
            result = update(
                tableName,
                values,
                specification.selection,
                specification.selectionArgs
            )
        }

        return result
    }

    override fun query(specification: SqlSpecification): List<T> {
        val database = dbHelper.readableCoreDatabase

        database.query(
            specification.isDistinct,
            tableName,
            specification.columns,
            specification.selection,
            specification.selectionArgs,
            specification.groupBy,
            specification.having,
            specification.orderBy,
            specification.limit
        ).use { cursor -> return mapCursorToResultList(cursor) }


    }

    override fun remove(specification: SqlSpecification) {
        val database = dbHelper.writableCoreDatabase

        return database.inTransaction {
            delete(
                tableName,
                specification.selection,
                specification.selectionArgs
            )
        }
    }

    override fun isEmpty(): Boolean {
        val database = dbHelper.readableCoreDatabase
        database.rawQuery(
            "SELECT COUNT(*) FROM $tableName;",
            null
        ).use { cursor ->
            cursor.moveToFirst()
            val count = cursor.getInt(cursor.getColumnIndexOrThrow("COUNT(*)"))
            return count == 0
        }
    }

    private fun mapCursorToResultList(cursor: Cursor): List<T> {
        val result: MutableList<T> = mutableListOf()
        if (cursor.moveToFirst()) {
            while (!cursor.isAfterLast) {
                val item = itemFromCursor(cursor)
                if (item != null) {
                    result.add(item)
                }
                cursor.moveToNext()
            }
        }
        return result
    }
}